benchmarks/BoltzmannWealth/boltzmann_wealth.py

```diff
@@ -21,18 +22,18 @@ class BoltzmannWealth(mesa.Model):
     def __init__(self, seed=None, n=100, width=10, height=10):
         super().__init__()
         self.num_agents = n
-        self.grid = mesa.space.MultiGrid(width, height, True)
+        self.grid = spaces.OrthogonalMooreGrid((width, height))
```

Comment:
Don't have a solution just yet, bit on first glance I don't directly like `OrthogonalMooreGrid` and `OrthogonalVonNeumannGrid`.

Maybe `OrthogonalGrid` with an required (so no default) `diagonal_connections` boolean keyword argument might work. That would also make it easier to experiment with it: Instead of needing to import another space, you switch a keyword argument.

Reply:
We debated this endlessly with the original PR. I don't want to redo that debate here. Let's first try and move all this code from experimental and take it from there in future PRs

Reply:
The problem is that once it's stable we can't as easily do things like this.

One part of stabilizing an API is re-evaluating properties of it. I know that will add some work, but if we're not doing that, then why did we make it experimental at all?

Reply:
Still you should start by revisiting the original PR. 

Regarding experimentation, I don't see a lot of problem changing `spaces.OrthogonalMooreGrid` to `spaces.OrthogonalVonNeumannGrid` If you import the spaces, no need to change imports.

Reply:
To be clear, I am happy to debate the API, but there is enough stuff going on in this PR that follows from resolving all the implications that follow from moving this from experimental to stable. I feel that adding the actual API to that long list adds further confusion. 

Moreover, we had good reasons in the original PR for the split and like @Corvince, I don't see a major practical difference between changing imports or a boolean.

Reply:
I actually can't find back significant debate about this split. The only things I can find where this comment that it was done (https://github.com/projectmesa/mesa/pull/1994#issuecomment-1950313918) and the one comment below being fine with it.

If `neighbord_func` is the bottleneck, that could just be a single `if diagonal_connections: ... else: ...` structure. It might even simplify the code, since other functions don't need to be duplicate anymore.

My main problem is that `Moore` and `VonNeumann` are two very specific names/concepts you have to remember, which actually convey very little information, while `diagonal_connections` (or `diagonals`) explains the concept directly by name and is thus intuitive.

Reply:
1. Moore is used as a label in the current version of the space but with a boolean. So we don't add new concepts to MESA.
2. Knowing Moore vs von Neumann neighborhoods in my view is basic ABM knowledge that a user should just have. It sits in the same category of knowing what a torus is. 
3. Explaining the difference between Moore and von Neumann is presently already covered in the docstring. We might expand this a bit more so a user does not need to google it.

Reply:
> Moore is used as a label in the current version of the space but with a boolean. 

So why then at the least not keep it this way? Would make migrating also easier.

Reply:
You normally subclass when you have either:
- Different (conflicting) attributes
- Different methods, with some absent in other

With diagonal/notdiagonal this is not the case, as far as I understand. They have the same methods, return the same outputs (a collection of cells) and have the same attributes (neighbours, etc.). The only thing that's different is a single function implementation (afaik), which can easily be covered by an if-else.

So why do I like keyword arguments (in this case)? Because they are more scalable. If you add another two booleans (on how many agents are allowed, another neighbour configuration, etc.), you can have 8 configuration options. With subclasses, you need to have 8 subclasses.

Here's another idea: Why not have a keyword argument `neighbour_definition`. Can be "Moore", "VonNeuman", but also something else, like a custom map. Could help a lot when we want to implement:
- https://github.com/projectmesa/mesa/issues/1900

It just feels like subclassing limits us significantly for future extension, and isn't the best practice in this case.

<hr>

I'm not going to die on this hill (in this PR). Move forward with it without this if you want, I can follow up.

Reply:
I am happy to explore this further, but indeed, preferably separately. 

When to subclass is a deeply contested topic, judging by the various StackOverflow questions about this that have been closed as being too opinionated. We indeed use it here for different implementations of how cells are wired up inside the grid. 

But I would not say that the fact that the only difference is the wiring up method implies we should (note not could, because that is indeed possible) reduce this all to a list of options for some `neighbour_definition` keyword argument and have a long case match statement to cover all possible neighborhood definitions. From a classic OO point of view, a Moore grid, a von Neuman grid, a network, a hexgrid, or a voroinoi grid, are all discrete grids. Subclassing them while abiding by a common interface is, in fact, good design. The motivation for subclassing in this reasoning is that their behavior is different.

Reply:
Just a note on the API (again, outside the scope of this PR): The use of subclassed to explicitly donate which space is used does not prohibit a possibly more user-friendly API. We could also think about a `create_space()` function with completely different arguments that return our spaces, without users needing to instantiate them themselves.